datapath: Don't enforce ingress policies at overlay if endpoint routes are enabled

pchaigno opened the pull request:
This fixes bug https://github.com/cilium/cilium/issues/14657. See commits for details.

There's probably no point in backporting because to fully fix this issue we would also need to backport https://github.com/cilium/cilium/pull/22190.

Fixes: https://github.com/cilium/cilium/pull/13346.
Fixes: https://github.com/cilium/cilium/issues/14657.

```release-note
Fix bug that caused ingress policies to be enforced twice when running with tunneling and endpoint routes.
```

pchaigno committed:
datapath: Don't enforce policies at overlay if ep routes are enabled

When endpoint routes are enabled, we should enforce ingress policies at
the destination lxc interface, to which a BPF program will be attached.
Nevertheless, today, for packets coming from the overlay, we enforce
ingress policies twice, once at the e.g. cilium_vxlan interface and a
second time at the lxc device.

This is happening for two reasons:
  1. bpf_overlay is not aware of the endpoint routes settings so it
     doesn't even know that it's not responsible for enforcing ingress
     policies.
  2. We have a flag to force the enforcement of ingress policies at the
     source in this case. This flag exists for historic reasons that are
     not valid anymore.

A separate patch will fix the reason 2 above. This commit fixes reason 1
by telling bpf_overlay to *not* enforce ingress policies when endpoint
routes are enabled.

Note that we do not support the case where some endpoint have endpoint
routes enabled and others don't. If we did, additional logic would be
required.

Fixes: 3179a4773 ("datapath: Support enable-endpoint-routes with encapsulation")
Signed-off-by: Paul Chaignon <paul@cilium.io>

julianwiedmann commented on pkg/datapath/linux/config/config.go:
```diff
@@ -770,6 +770,10 @@ func (h *HeaderfileWriter) writeNetdevConfig(w io.Writer, cfg datapath.DeviceCon
 		}
 		fmt.Fprint(w, "\n")
 	}
+
+	if option.Config.EnableEndpointRoutes {
+		fmt.Fprint(w, "#define USE_BPF_PROG_FOR_INGRESS_POLICY 1\n")
+	}
```
> Note that we do not support the case where some endpoint have endpoint
> routes enabled and others don't. If we did, additional logic would be
> required.

I wonder if we should make that clearer in the code, and place the above code into `WriteNodeConfig()` (removing the `e.RequireEgressProg` [case](https://github.com/cilium/cilium/blob/a4470123018163288e2a0f040856ede1319f27d7/pkg/datapath/linux/config/config.go#L866-L868) in `writeTemplateConfig()` altogether).

pchaigno replied:
That's a good point. I'm however a bit wary of making that change because it's more likely to break someone's setup (running in chaining mode without endpoint routes maybe). Even if we don't want to support that, I'd rather avoid breaking someone's setup outright. I'm in favor of making that change but I'll bring it up for discussion first.

The case of cilium_vxlan -> pod is a bit different because (1) it's already buggy and (2) it's a very arcane setup we've never publicly recommended AFAIK.

EDIT: I'm not backporting this for now, but I suspect we may need to. So another reason to keep the change limited in scope.

julianwiedmann replied:
>Even if we don't want to support that, I'd rather avoid breaking someone's setup outright. I'm in favor of making that change but I'll bring it up for discussion first.

Ack, it's worth its own PR and corresponding doc update.

pchaigno replied:
cc @aditighag I think that answers your question.

nbusseneau reviewed (approved):
Auto-acking for ci-structure, trivial changes.

aditighag reviewed (commented):
I haven't checked out the PR, and tested yet. But IIRC, service rev translation happens in the overlay program, so with this fix, won't we end up enforcing an ingress policy on the translated address? Did you manually test this fix for pod -> svc ip (selected backend on remote node) + policy case with the tunneling and ep routes combination? (I see that tests have passed, but I'm not sure about the test coverage for this case.)

aditighag reviewed (approved):
> > But IIRC, service rev translation happens in the overlay program, so with this fix, won't we end up enforcing an ingress policy on the translated address?
> 
> No. The reverse translation happens before the local delivery. See `handle_ipv4` in `bpf_overlay.c`.

👍 You are right. Any chance we can combine this with the per endpoint check - https://github.com/cilium/cilium/blob/master/pkg/datapath/linux/config/config.go#L862-L862? This is needed in general whenever endpoint routes are enabled, no?

pchaigno committed:
bpf: Preserve overlay->lxc path with kube-proxy

The previous commit changed the packet handling on the path
overlay->lxc to fix a bug. More presicely, when endpoint routes are
enabled, we won't enforce ingress policies on both the overlay and the
lxc devices but only on the latter.

However, as a consequence of that patch, we don't go through the
policy-only program in bpf_lxc and we therefore changed the way the
packet is transmitted between overlay and lxc devices in some cases. As
a summary of changes made in the previous path, consider the following
table for the path overlay -> lxc.

Before the previous patch:
| Endpoint routes | Enforcement     | Path                 |
|-----------------|-----------------|----------------------|
| Enable          | overlay AND lxc | bpf_redirect if KPR; |
|                 |                 | stack otherwise      |
| Disabled        | overlay         | bpf_redirect         |

Now:
| Endpoint routes | Enforcement | Path         |
|-----------------|-------------|--------------|
| Enable          | lxc         | bpf_redirect |
| Disabled        | overlay     | bpf_redirect |

The previous patch intended to fix the enforcement to avoid the double
policy enforcement, but it also changed the packet path in case endpoint
routes are enabled.

This patch now fixes this by adding the same exception we have in
bpf_lxc to the l3.h logic we have. Hence, with the current patch, the
table will look like:
| Endpoint routes | Enforcement | Path                 |
|-----------------|-------------|----------------------|
| Enable          | lxc         | bpf_redirect if KPR; |
|                 |             | stack otherwise      |
| Disabled        | overlay     | bpf_redirect         |

I've kept this in a separate commit from the previous in an attempt to
split up and the logic and more clearly show the deltas.

Signed-off-by: Paul Chaignon <paul@cilium.io>
